src/nn/residual.rs

```diff
@@ -1,52 +1,105 @@
 use crate::prelude::*;
 
-/// A residual connection around `F`: `F(x) + x`,
+/// A residual connection `R` around `F`: `F(x) + R(x)`,
 /// as introduced in [Deep Residual Learning for Image Recognition](https://arxiv.org/abs/1512.03385).
 ///
 /// # Generics
 /// - `F`: The underlying module to do a skip connection around.
+/// - `R`: The underlying residual module
 ///
 /// # Examples
 /// ```rust
 /// # use dfdx::prelude::*;
-/// let module: Residual<ReLU> = Default::default();
+/// let module: Residual<ReLU, Square> = Default::default();
 /// let x = Tensor1D::new([-2.0, -1.0, 0.0, 1.0, 2.0]);
 /// let y = module.forward(x);
-/// assert_eq!(y.data(), &[-2.0, -1.0, 0.0, 2.0, 4.0]);
+/// assert_eq!(y.data(), &[4.0, 1.0, 0.0, 2.0, 6.0]);
 /// ```
 #[derive(Debug, Clone, Default)]
-pub struct Residual<F>(F);
+pub struct Residual<F, R>(F, R);
 
-impl<F: CanUpdateWithGradients> CanUpdateWithGradients for Residual<F> {
+impl<F: CanUpdateWithGradients, R: CanUpdateWithGradients> CanUpdateWithGradients
+    for Residual<F, R>
+{
     /// Pass through to `F`'s [CanUpdateWithGradients].
     fn update<G: GradientProvider>(&mut self, grads: &mut G) {
         self.0.update(grads);
+        self.1.update(grads);
     }
 }
 
-impl<F: ResetParams> ResetParams for Residual<F> {
+impl<F: ResetParams, R: ResetParams> ResetParams for Residual<F, R> {
     /// Pass through to `F`'s [ResetParams].
-    fn reset_params<R: rand::Rng>(&mut self, rng: &mut R) {
+    fn reset_params<RNG: rand::Rng>(&mut self, rng: &mut RNG) {
         self.0.reset_params(rng);
+        self.1.reset_params(rng);
     }
 }
 
-impl<T, F> Module<T> for Residual<F>
+impl<O, T, F, R> Module<T> for Residual<F, R>
 where
-    T: Tensor<Dtype = f32>,
-    F: Module<T, Output = T>,
+    O: Tensor<Dtype = f32>,
+    T: Tensor<Dtype = f32, Tape = O::Tape>,
+    F: Module<T, Output = O>,
+    R: Module<T, Output = O>,
 {
-    type Output = F::Output;
+    type Output = O;
 
-    /// Calls forward on `F` and then adds `x` to the result: `F(x) + x`
+    /// Calls forward on `F` and `R` and then sums their result: `F(x) + R(x)`
     fn forward(&self, x: T) -> Self::Output {
```

Comment:
I'm thinking this forward function can be simplified a lot with something like:
```rust
impl<F, R, T> Module<T> for Residual<F, R> {
where
    T: Tensor<Dtype = f32>,
    F: Module<T>,
    R: Module<T, Output = F::Output>,
{
    type Output = F::Output;
    fn forward(&self, x: T) -> Self:: Output {
        let (x, tape) = x.split_tape();

        // do R(x) on the tape
        let (r_x, tape) = self.1.forward(x.duplicate().put_tape(tape)).split_tape();

        // do F(x) on the tape
        let f_x = self.0.forward(x.put_tape(tape));

        add(f_x, &r_x)
    }
}
```

Reply:
ok, makes sense. i think that i am missing lots of differentiation basics :/

Reply:
No worries. So far the modules in nn don't actually specify any backward ops, they just either call other module's forwards or call operations from tensor_ops.